### ISSUE TITLE
Please consider providing Graph Reachability Matrix predicate

### ISSUE THREAD

**Issue by LebedevRI** [MiniZinc/libminizinc#837]
It would have been rather useful to me when i was trying to write a model,
and i couldn't quite come up with anything good on a spot,
and thus after much pain ended up with: 
https://github.com/LebedevRI/minizinc-graph-reachability-matrix-predicate.mzn

Unfortunately it ends up having `O(n^3)` complexity, as benchmark notes,
but that is only because `connected()` is itself `O(n^2)`.
There may perhaps be a better formulation for all that,
but i'm not seeing it currently. 
The results are already much better than i had hoped.

(There may be a standard set of inefficiencies in minizinc itself,
not sure how much i want to look into that given how it's developed and tested...)

**Comment by LebedevRI** [comment]
FYI i'm looking at a non-cubic implementation, hang on.

**Comment by LebedevRI** [comment]
@cyderize @guidotack Ok! I think i got what i was looking for. 

Essentially, this ended up consisting of two parts:
1. `reachable_multiroot()` - a seemingly rather boringly straight-forward generalization of the `reachable()`,
    which, instead of enforcing that the subgraph is reachable from _the_ root node (given by index),
    and thus forcing there to be a single joint subgraph, instead takes (zero or more) roots as a mask.
    The rest of the logic is generalized accordingly.

```
include "subgraph.mzn";

predicate fzn_dreachable(array[int] of $$N: from, array[int] of $$N: to,
                        var $$N: r, array[$$N] of var bool: ns, array[int] of var bool: es) =
    let { 
        array[index_set(ns)] of var 0..card(index_set(ns))-1: dist; /* distance from root */
        array[index_set(ns)] of var index_set(ns): parent; /* parent */
    } in
        ns[r] /\ % the root must be chosen
        dist[r] = 0 /\ % root is at distance 0
        parent[r] = r /\ % root is its own parent
        forall(n in index_set(ns)) % nonselected nodes have parent 0
              (not ns[n] -> parent[n] = n) /\
        forall(n in index_set(ns)) % nonselected nodes have distance 0
              (not ns[n] -> dist[n] = 0) /\
        forall(n in index_set(ns)) % each in node except root must have a parent
              (ns[n] -> (n = r \/ parent[n] != n)) /\
        forall(n in index_set(ns)) % each in node with a parent must be in and also its parent
              (parent[n] != n -> (ns[n] /\ ns[parent[n]])) /\
        forall(n in index_set(ns)) % each except with a parent is one more than its parent
              (parent[n] != n -> dist[n] = dist[parent[n]] + 1) /\
        forall(n in index_set(ns)) % each node with a parent must have that edge in
              (parent[n] != n -> exists(e in index_set(from) where to[e] = n)(es[e] /\ from[e] = parent[n])) /\
        subgraph(from,to,ns,es);
   
%-----------------------------------------------------------------------------%
```
vs
```
include "subgraph.mzn";

predicate fzn_dreachable_multiroot(array[$$E] of $$N: from, array[$$E] of $$N: to,
                                   array[$$N] of var bool: rs, array[$$N] of var bool: ns, array[$$E] of var bool: es) =
    let {
        array[index_set(ns)] of var 0..card(index_set(ns))-1: dist; /* distance from root */
        array[index_set(ns)] of var index_set(ns): parent; /* parent */
    } in
        forall(n in index_set(ns)) % root nodes must be chosen
              (rs[n] -> ns[n]) /\
        forall(n in index_set(ns)) % self-parent is the same as zero distance from root
              ((parent[n] = n) == (dist[n] = 0)) /\
        forall(n in index_set(ns)) % self-parent is the same as either non-chosen node or root node
              ((parent[n] = n) == ((not ns[n]) \/ rs[n])) /\
        forall(n in index_set(ns)) % each non-self-parent node is one more step removed from a root than its parent node
              (parent[n] != n -> dist[n] = dist[parent[n]] + 1) /\
        forall(n in index_set(ns)) % each non-self-parent node must have a chosen edge from its parent
              (parent[n] != n -> exists(e in index_set(from) where to[e] = n)(es[e] /\ from[e] = parent[n])) /\
        % nodes connected by chosen edges must be chosen
        subgraph(from,to,ns,es);

%-----------------------------------------------------------------------------%
```

2. `cluster()` predicate.
    This one took **_quite_** a bit of rewrites to fully distill it to it's purest+simplest+fastest form. :)
    The (final) basic idea is to, basically, give each node a unique "color",
    and then set the nodes, linked together by selected edges, to the same color,
    thus ensuring that the nodes, that are not in any way connected, don't have the same "color",
    and vice versa.

    Implementation-wise, we just defer to the `reachable_multiroot()` to pick the root nodes
    of disjoint subgraphs, and then compute reachability matrix based on whether or not
    two nodes belong to the same disjoint subgraph (= have same root).

    **Now, this is controversial. I've decided to return the reachability matrix,
    *NOT* the subgraph index. While sparse data structure seems better,
    it immediately requires symmetry breaking constraint, which is tricky (i have written it.),
    and `--all-solutions` actually ends up being *MUCH* slower (`O^3`?)
    than just returning reachability matrix...
    
```
include "link_set_to_booleans.mzn";
include "reachable_multiroot.mzn";
include "subgraph.mzn";
include "arg_val.mzn";

predicate fzn_cluster(array[$$E] of $$N: from, array[$$E] of $$N: to,
                      array[$$N,$$N] of var bool: r,
                      array[$$N] of var bool: ns, array[$$E] of var bool: es) =
    let {
        array[index_set(ns)] of var index_set(ns): NodeRoot;
        array[index_set(ns)] of var bool: NodeRootnessMask;
        array[index_set(ns)] of bool: ALL_GRAPH_NODES = [ true | n in index_set(ns) ];
    } in
        forall (n in index_set(ns))(
            ((not ns[n]) -> (NodeRoot[n] == n))
        ) /\
        forall (e in index_set(es))(
            es[e] -> (NodeRoot[from[e]] == NodeRoot[to[e]])
        ) /\
        link_set_to_booleans(array2set(NodeRoot), NodeRootnessMask) /\
        reachable_multiroot(
            from,
            to,
            NodeRootnessMask,
            ALL_GRAPH_NODES,
            es
        ) /\
        symmetry_breaking_constraint(forall (j,i in index_set(ns) where i > j)(
            r[i,j] == r[j,i]
        )) /\
        forall (c in index_set(ns))(
            r[c,c] == true
        ) /\
        forall (j,i in index_set(ns) where i > j)(
            r[j,i] == (NodeRoot[i] == NodeRoot[j])
        );
```

I have an exhaustive test harness for the `cluster()` (but not the rest),
it would not have been possible to refine the predicate so much otherwise,
please see https://github.com/LebedevRI/minizinc-graph-reachability-matrix-predicate.mzn
[minizinc-graph-reachability-matrix-predicate.zip](https://github.com/user-attachments/files/18722539/minizinc-graph-reachability-matrix-predicate.zip)

Knowing that this repo is just a mirror, and PR would not be merged anyways,
what else would you want to see before this can be merged?
Please do ensure that the commit does end up being committed with `--author="Roman Lebedev <lebedev.ri@gmail.com>"` to ensure proper attribution.

**Comment by LebedevRI** [comment]
Just out of interest, naive directional `dcluster` seems to have ended up being being somewhat straight-forward,
(i mean, at least after spending all that time coming up with unidirectional `cluster`...)
not really sure yet how much it can be improved
https://github.com/LebedevRI/minizinc-graph-reachability-matrix-predicate.mzn/blob/proof-of-concept-directional/include/fzn_dcluster_enum.mzn
```
predicate fzn_dcluster(array[$$E] of $$N: from, array[$$E] of $$N: to,
                       array[$$N,$$N] of var bool: r,
                       array[$$N] of var bool: ns, array[$$E] of var bool: es) =
    let {
        array [index_set(ns),index_set(es)] of var bool: des,
    } in
        forall (c in index_set(ns))(
            r[c,c] == true
        ) /\
        forall (j,i in index_set(ns) where i != j)(
            r[j,i] -> (ns[j] /\ ns[i])
        ) /\
        forall (j,i in index_set(ns) where i != j)(
            r[j,i] == (
                exists(e in index_set(es) where from[e] == j /\ to[e] == i)(
                    es[e]
                ) \/
                exists(k in index_set(ns) where k != j /\ k != i)(
                    ns[k] /\ r[j,k] /\ r[k,i]
                )
            )
        ) /\
        forall (e in index_set(es))(
            exists(des[..,e]) -> es[e]
        ) /\
        forall (j in index_set(ns))(
            dreachable(from, to, j, r[j,..], array1d(index_set(es),des[j,..]))
        ) /\
        % nodes connected by chosen edges must be chosen
        subgraph(from,to,ns,es);

```

**Comment by LebedevRI** [comment]
> not really sure yet how much it can be improved

Relevant papers:
* Luis Quesada and Kenneth N. Brown. Positive and Negative Length-Bound Reachability Constraints. In 27th International Conference on Principles and Practice of Constraint Programming (CP 2021). Leibniz International Proceedings in Informatics (LIPIcs), Volume 210, pp. 46:1-46:16, Schloss Dagstuhl – Leibniz-Zentrum für Informatik (2021) https://doi.org/10.4230/LIPIcs.CP.2021.46
* Using Dominators for Solving Constrained Path Problems
   Luis Quesada, Peter Van Roy, Yves Deville, and Rapha¨el Collet
   https://webperso.info.ucl.ac.be/~pvr/padl06cr.pdf
* Solving Constrained Graph Problems using Reachability Constraints based on Transitive Closure and Dominators
   Luis Quesada
   https://webperso.info.ucl.ac.be/~yde/Papers/thesis-quesada2006.pdf

TLDR is: the approach i happened to find (using `dreachable`) is apparently second best,
to improve upon that one needs to (efficiently) model dominators via constraints...

**Comment by LebedevRI** [comment]
So i've finally wanted to try to integrate it into the libminizinc,
and i'm confused as to how testsuite operates.
Is it not testing the `share/minizinc/std/` from the source tree,
but the default global installed one?
